viewer/packages/camera-manager/src/CameraManager.ts

```diff
@@ -0,0 +1,584 @@
+      this._onClick = undefined;
+    }
+    if (this._onWheel !== undefined) {
+      this._domElement.removeEventListener('wheel', this._onWheel);
+      this._onWheel = undefined;
+    }
+  }
+  private handleMouseWheelActionChange(controlsOptions: CameraControlsOptions) {
+    const { _controls: controls } = this;
+
+    switch (controlsOptions?.mouseWheelAction) {
+      case 'zoomToTarget':
+        controls.zoomToCursor = false;
+        break;
+      case 'zoomPastCursor':
+        controls.useScrollTarget = false;
+        controls.zoomToCursor = true;
+        break;
+      case 'zoomToCursor':
+        controls.setScrollTarget(controls.getState().target);
+        controls.useScrollTarget = true;
+        controls.zoomToCursor = true;
+        break;
+      case undefined:
+        break;
+      default:
+        assertNever(controlsOptions.mouseWheelAction);
+    }
+  }
+  /**
+   * Method for setting up camera controls listeners and values inside current controls class.
+   */
+  private setupControls() {
+    let scrollStarted = false;
+
+    const wheelClock = new THREE.Clock();
+
+    const onClick = async (e: any) => {
+      this._controls.enableKeyboardNavigation = false;
+      const newTarget = await this.calculateNewTarget(e);
+      this.setCameraTarget(newTarget, true);
+    };
+
+    const onWheel = async (e: any) => {
+      const timeDelta = wheelClock.getDelta();
+
+      const wantNewScrollTarget = !scrollStarted && e.deltaY < 0;
+
+      if (wantNewScrollTarget) {
+        scrollStarted = true;
+
+        const newTarget = await this.calculateNewTarget(e);
+        this._controls.setScrollTarget(newTarget);
+      } else if (timeDelta > 0.08) {
+        scrollStarted = false;
+      }
+    };
+
+    this.handleMouseWheelActionChange(this._cameraControlsOptions);
+
+    if (this._cameraControlsOptions.changeCameraTargetOnClick) {
+      this._inputHandler.on('click', onClick);
+      this._onClick = onClick;
+    }
+    if (this._cameraControlsOptions.mouseWheelAction === 'zoomToCursor') {
+      this._domElement.addEventListener('wheel', onWheel);
+      this._onWheel = onWheel;
+    }
+  }
+
+  /**
+   * Convert pixel coordinates of the cursor to [-1,1]^2 coordinates.
+   * @param pixelX
+   * @param pixelY
+   */
+  private convertPixelCoordinatesToNormalized(pixelX: number, pixelY: number) {
+    const x = (pixelX / this._domElement.clientWidth) * 2 - 1;
+    const y = (pixelY / this._domElement.clientHeight) * -2 + 1;
+
+    return { x, y };
+  }
+
+  private calculateDefaultDuration(distanceToCamera: number): number {
+    let duration = distanceToCamera * 125; // 125ms per unit distance
+    duration = Math.min(
+      Math.max(duration, CameraManager.DefaultMinAnimationDuration),
+      CameraManager.DefaultMaxAnimationDuration
+    );
+
+    return duration;
+  }
+}
+
+function getBoundingBoxCorners(bbox: THREE.Box3, outBuffer?: THREE.Vector3[]): THREE.Vector3[] {
```

Comment:
Can this be replaced with `visitBox3CornerPoints` which we have in the utilities-package?

Reply:
I checked `getBox3CornerPoints` method and it is essentially doing the same thing (only the order of corners is different but it shouldn't matter in most (all?) cases) but every function call it is creating a new array. However, implementation in `getBoundingBoxCorners` work on present array if it is passed. Since call is happening every frame we need to stick to last implementation. We can try just to remove method from CameraManager and change implementation in utils if it is ok.

Reply:
Maybe you at least can implement getBox3CornerPoints using visitBox3CornerPoints?

Reply:
Not sure what you mean here, there is no method called `visitBox3CornerPoints`, you meant `getBox3CornerPoints`? If yes, then you suggest using it for `getBoundingBoxCorners` or vice versa?

Reply:
Not sure the exact naming but there was a function in the utilities package I thought could be used here to implement the function

Reply:
Anyway, minor. It's OK as is.